File: Problem/Easy/125. ValidPalindrome.py
# Brute Force Solution
class Solution:
    def isPalindrome(self, s: str) -> bool:
        new_string=[character for character in s if character.isalnum()]
        new_string = "".join(new_string).lower()

        length = len(new_string)
        reverse =new_string[length::-1]

        if(reverse == new_string):
            return True
        else:
            return False

# Efficient Solution    
class Solution:
    def isPalindrome(self, s: str) -> bool:
	# upper case convert lower case
	# delete punctuation and space
	nor_str = ''.join([c.lower() if c.isalnum() else '' for c in s])
	# two pinters check
    return nor_str[:] == nor_str[::-1]

File: Problems/Easy/125. Valid Palindrome.java
// Optimal Solution
public class Solution {
    public boolean isPalindrome(String s) {
        int start = 0; 
        int end = s.length() - 1;
        while(start <= end) {
            while(start <= end && !Character.isLetterOrDigit(s.charAt(start))) {
                start++;
            }
            while(start <= end && !Character.isLetterOrDigit(s.charAt(end))) {
                end--;
            }
            if(start <= end && Character.toLowerCase(s.charAt(start)) != Character.toLowerCase(s.charAt(end))) {
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
}

// Brute Force
class Solution {
    public boolean isPalindrome(String s) {
        s = s.replaceAll("[^a-zA-Z0-9]", "").toLowerCase();
        
        int i = 0, j = s.length()-1;
        
            while(j>i){
                if(s.charAt(i) != s.charAt(j)){
                    return false;
                }
                else{
                    i++;
                    j--;
                }
            }
         return true;
    }
}